### PR TITLE
Panwang/obcache

### PR DESCRIPTION
**Pull request by pan-wang**
cache the handler object to avoid potential heap fragmenting issue and fix some logging and AV issue

### CODE REVIEW

File: src/AspNetCoreModuleV2/InProcessRequestHandler/inprocesshandler.cpp [hunk 1]
```diff
@@ -59,6 +61,80 @@ IN_PROCESS_HANDLER::OnExecuteRequestHandler()
...
+        }
+
+        hr = pAlloc->Initialize(sizeof(IN_PROCESS_HANDLER),
+            64); // nThreshold
+        if (FAILED(hr))
+        {
+            goto Failure;
+        }
+
+        AcquireSRWLockExclusive(&g_srwLockRH);
```
Comment (pakrym) [inline src/AspNetCoreModuleV2/InProcessRequestHandler/inprocesshandler.cpp]: Why the lock?

File: src/AspNetCoreModuleV2/InProcessRequestHandler/inprocesshandler.cpp [hunk 2]
```diff
@@ -59,6 +61,65 @@ IN_PROCESS_HANDLER::OnExecuteRequestHandler()
...
+    DBG_ASSERT(sm_pAlloc != NULL);
+    if (sm_pAlloc != NULL)
+    {
+        sm_pAlloc->Free(pMemory);
+    }
+}
+
+// static
+HRESULT
+IN_PROCESS_HANDLER::StaticInitialize( VOID )
```
Comment (jkotalik) [inline src/AspNetCoreModuleV2/InProcessRequestHandler/inprocesshandler.cpp]: I don't like the static initialize pattern; it's easy to forget calling it. Calling it in dll main also seems odd. Can we put this in the inprocess application?

File: src/AspNetCoreModuleV2/OutOfProcessRequestHandler/outofprocess/forwardinghandler.h [hunk 3]
```diff
@@ -72,6 +72,10 @@ class FORWARDING_HANDLER : public REQUEST_HANDLER
         bool    fClientInitiated
     );
 
+    static void * operator new(size_t size);
```
Comment (jkotalik) [inline src/AspNetCoreModuleV2/OutOfProcessRequestHandler/outofprocess/forwardinghandler.h]: override?

**Comment by pakrym** [comment]
I have another idea to address startup failures, let's create another IAPPLICATION implementation that would return 502 and return it in startup error cases, this would allow us to move even more code to handlers. Also in this case restart just works, because when configuration would change application would get restarted without requiring additional logic.

File: src/AspNetCoreModuleV2/InProcessRequestHandler/inprocessapplication.cpp [hunk 4]
```diff
@@ -659,17 +659,20 @@ IN_PROCESS_APPLICATION::LogErrorsOnMainExit(
...
-            ASPNETCORE_EVENT_INPROCESS_THREAD_EXIT_STDOUT_MSG,
-            m_pConfig->QueryApplicationPath()->QueryStr(),
-            m_pConfig->QueryApplicationPhysicalPath()->QueryStr(),
-            hr,
-            struStdErrOutput.QueryStr());
+    STRA straStdErrOutput;
+    STRU struStdMsg;
+    if (m_pLoggerProvider->GetStdOutContent(&straStdErrOutput))
+    {
+        if (SUCCEEDED(struStdMsg.CopyA(straStdErrOutput.QueryStr()))) {
```
[the remark this answers is not in the document]
Comment (pakrym) [reply src/AspNetCoreModuleV2/InProcessRequestHandler/inprocessapplication.cpp]: Why is stdout content CHAR*?

File: src/AspNetCoreModuleV2/InProcessRequestHandler/inprocesshandler.cpp [hunk 5]
```diff
@@ -59,6 +61,65 @@ IN_PROCESS_HANDLER::OnExecuteRequestHandler()
     return m_pApplication->OnExecuteRequest(m_pW3Context, this);
 }
 
+// static
+void * IN_PROCESS_HANDLER::operator new(size_t)
+{
+    DBG_ASSERT(sm_pAlloc != NULL);
+    if (sm_pAlloc == NULL)
+    {
+        return NULL;
```
Comment (pakrym) [inline src/AspNetCoreModuleV2/InProcessRequestHandler/inprocesshandler.cpp]: Not a fan of returning null, I'd like to stop to check allocations for NULL every time. Can you throw bad_alloc here please?

File: src/AspNetCoreModuleV2/InProcessRequestHandler/inprocesshandler.cpp [hunk 6]
```diff
@@ -59,6 +61,65 @@ IN_PROCESS_HANDLER::OnExecuteRequestHandler()
...
+    HRESULT                         hr = S_OK;
+
+    sm_pAlloc = new ALLOC_CACHE_HANDLER;
+    if (sm_pAlloc == NULL)
+    {
+        hr = E_OUTOFMEMORY;
+        goto Finished;
+    }
+
+    hr = sm_pAlloc->Initialize(sizeof(IN_PROCESS_HANDLER),
```
[the remark this answers is not in the document]
Comment (pan-wang) [reply src/AspNetCoreModuleV2/InProcessRequestHandler/inprocesshandler.cpp]: yes

File: src/AspNetCoreModuleV2/InProcessRequestHandler/inprocesshandler.cpp (hunk 2 above)
Comment (pan-wang) [reply src/AspNetCoreModuleV2/InProcessRequestHandler/inprocesshandler.cpp]: the sm_pAlloc is a static variable shared cross applications in the same worker process. We have to call some static initialize function to init it.

File: src/AspNetCoreModuleV2/InProcessRequestHandler/inprocesshandler.cpp (hunk 5 above)
Comment (pan-wang) [reply src/AspNetCoreModuleV2/InProcessRequestHandler/inprocesshandler.cpp]: yes, you can to try-catch on bad_alloc exception.

Comment (pakrym) [reply src/AspNetCoreModuleV2/InProcessRequestHandler/inprocesshandler.cpp]: Not, what I meant is can you throw bad_alloc if sm_pAlloc  is null, instead of returning null?

File: src/AspNetCoreModuleV2/InProcessRequestHandler/inprocessapplication.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (pan-wang) [reply src/AspNetCoreModuleV2/InProcessRequestHandler/inprocessapplication.cpp]: i don't want to touch the original design for pipeoutputmanage which passed a char buffer to receive the output. We can change the buffere to wchar.

File: src/AspNetCoreModuleV2/AspNetCore/src/applicationinfo.cpp [hunk 7]
```diff
@@ -246,6 +252,20 @@ APPLICATION_INFO::EnsureApplicationCreated(
...
+    {
+        // Log the failure and update application info to not try again
+        UTILITY::LogEventF(g_hEventLog,
+            EVENTLOG_ERROR_TYPE,
+            ASPNETCORE_EVENT_ADD_APPLICATION_ERROR,
+            ASPNETCORE_EVENT_ADD_APPLICATION_ERROR_MSG,
+            pHttpContext->GetApplication()->GetApplicationId(),
+            hr);
+
+        goto Finished;
```
[the remark this answers is not in the document]
Comment (pan-wang) [reply src/AspNetCoreModuleV2/AspNetCore/src/applicationinfo.cpp]: fixed

File: src/AspNetCoreModuleV2/AspNetCore/Inc/applicationinfo.h [hunk 8]
```diff
@@ -150,6 +156,8 @@ class APPLICATION_INFO
     STRU                    m_struInfoKey;
     BOOL                    m_fAppOfflineFound;
     BOOL                    m_fAllowStart; // Flag indicates whether there is (configuration) error blocking application from starting
+    BOOL                    m_fAppCreated; // Flag indicates if the application has been created successfully (runtime error).
```
[the remark this answers is not in the document]
Comment (pan-wang) [reply src/AspNetCoreModuleV2/AspNetCore/Inc/applicationinfo.h]: yes, we can resue APPLICATION_STATUS

File: src/AspNetCoreModuleV2/InProcessRequestHandler/inprocesshandler.cpp (hunk 5 above)
Comment (pan-wang) [reply src/AspNetCoreModuleV2/InProcessRequestHandler/inprocesshandler.cpp]: for your comments about moving check createapplication failure to requesthandler. it is possible. It requires change for inprocess. Will do it with moving app_offline

File: src/AspNetCoreModuleV2/AspNetCore/Inc/applicationinfo.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (jkotalik) [reply src/AspNetCoreModuleV2/AspNetCore/Inc/applicationinfo.h]: That's not what I intended to ask. I just think if we are allowed to start, it means the app is created.

File: src/AspNetCoreModuleV2/AspNetCore/src/applicationmanager.cxx [hunk 9]
```diff
@@ -254,25 +253,9 @@ APPLICATION_MANAGER::RecycleApplicationFromManager(
...
-            delete m_pApplicationInfoHash;
-
-            // We don't want to delete the table as m_pApplicationInfoHash = table
-            fKeepTable = TRUE;
-            m_pApplicationInfoHash = table;
-        }
-    }
-
-    if (m_pApplicationInfoHash->Count() == 0)
+    if (m_pApplicationInfoHash->Count() == 0 && m_hostingModel == HOSTING_OUT_PROCESS)
```
Comment (jkotalik) [inline src/AspNetCoreModuleV2/AspNetCore/src/applicationmanager.cxx]: Why was this changed?
